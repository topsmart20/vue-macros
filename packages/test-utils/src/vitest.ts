import path from 'node:path'
import glob, { type Options as GlobOptions } from 'fast-glob'
import { describe, expect, test } from 'vitest'
import { normalizePath } from '@rollup/pluginutils'

interface Options {
  params?: [name: string, values?: any[]][]
  promise?: boolean
}

const SKIP_VUE2 = !!process.env.SKIP_VUE2

export async function testFixtures(
  globs: string | string[],
  exec: (args: Record<string, any>, id: string) => any,
  options?: GlobOptions & Options,
): Promise<void>
export async function testFixtures(
  files: Record<string, string>,
  exec: (args: Record<string, any>, id: string, code: string) => any,
  options?: Options,
): Promise<void>
export async function testFixtures(
  globsOrFiles: string | string[] | Record<string, string>,
  cb: (args: Record<string, any>, id: string, code: string) => any,
  { params, promise, ...globOptions }: GlobOptions & Options = {},
) {
  let files: Record<string, string | undefined>

  if (typeof globsOrFiles === 'string' || Array.isArray(globsOrFiles)) {
    files = Object.fromEntries(
      (await glob(globsOrFiles, globOptions)).map((file) => [file, undefined]),
    )
  } else {
    files = globsOrFiles
  }

  for (const [id, code] of Object.entries(files)) {
    makeTests(id, code, [[normalizePath(id)], ...(params || [])])()
  }

  function getName(name: string, value: any) {
    return value !== undefined ? `${name} = ${String(value)}` : name
  }

  function makeTests(
    id: string,
    code: string | undefined,
    params: NonNullable<Options['params']>,
    args: Record<string, any> = {},
  ) {
    const [currParams, ...restParams] = params
    const [name, values = [undefined]] = currParams
    if (restParams.length > 0) {
      return () => {
        for (const value of values) {
          const currArgs = { ...args, [name]: value }

          describe(
            getName(name, value),
            makeTests(id, code, restParams, currArgs),
          )
        }
      }
    } else {
      return () => {
        for (const value of values) {
          const testName = getName(name, value)
          const isSkip = testName.includes('vue2') && SKIP_VUE2

          test.skipIf(isSkip)(testName, async () => {
            const currArgs = { ...args, [name]: value }
            const execute = () =>
              cb(currArgs, path.resolve(globOptions.cwd || '/', id), code!)
            if (id.includes('error')) {
              if (promise) {
                await expect(execute()).rejects.toThrowErrorMatchingSnapshot()
              } else {
                expect(execute).toThrowErrorMatchingSnapshot()
              }
            } else if (promise) {
              await expect(
                (execute() as Promise<any>).catch((error) => {
                  console.warn(error)
                  return Promise.reject(error)
                }),
              ).resolves.toMatchSnapshot()
            } else {
              expect(execute()).toMatchSnapshot()
            }
          })
        }
      }
    }
  }
}
