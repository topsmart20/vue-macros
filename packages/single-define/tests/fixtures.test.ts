import { resolve } from 'node:path'
import { describe } from 'vitest'
import {
  RollupEsbuildPlugin,
  RollupToStringPlugin,
  rollupBuild,
  testFixtures,
} from '@vue-macros/test-utils'
import VueSingleDefine from '../src/rollup'

describe('fixtures', async () => {
  await testFixtures(
    'tests/fixtures/*.vue',
    (args, id) =>
      rollupBuild(id, [
        VueSingleDefine({ isProduction: false }),
        RollupToStringPlugin(),
        RollupEsbuildPlugin({
          target: 'esnext',
        }),
      ]),
    { cwd: resolve(__dirname, '..'), promise: true }
  )
})
