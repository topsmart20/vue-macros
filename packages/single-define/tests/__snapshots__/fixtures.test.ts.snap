// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fixtures > tests/fixtures/define-emit.vue 1`] = `
"var defineEmit = \`<script setup lang=\\"ts\\">
const __MACROS_emit = defineEmits({
      \\"foo\\": null,
  \\"bar\\": (value: string) => typeof value === 'string',
  \\"baz\\": (value) => typeof value === 'number',
  \\"baz\\": (value) => typeof value === 'string',
  \\"change\\": null
    })

import { expectTypeOf } from 'expect-type'

const foo = (...args) => __MACROS_emit(\\"foo\\", ...args)
expectTypeOf(foo).toEqualTypeOf<(...args: any[]) => void>()

const bar = (...args) => __MACROS_emit(\\"bar\\", ...args)
expectTypeOf(bar).toEqualTypeOf<(value: string) => void>()

const baz = (...args) => __MACROS_emit(\\"baz\\", ...args)
expectTypeOf(baz).toEqualTypeOf<(value: number) => void>()

const qux = (...args) => __MACROS_emit(\\"baz\\", ...args)
expectTypeOf(qux).toEqualTypeOf<(value: string) => void>()

const change = (...args) => __MACROS_emit(\\"change\\", ...args)
expectTypeOf(change).toEqualTypeOf<() => void>()
</script>
\`;

export { defineEmit as default };
"
`;

exports[`fixtures > tests/fixtures/define-prop.vue 1`] = `
"var defineProp = \`<script setup lang=\\"ts\\">
const __MACROS_props = defineProps({
  \\"foo\\": { type: String },
  \\"bar\\": { ...{ type: String }, ...{
  type: String,
  required: true,
  default: 'bar',
} },
  \\"baz\\": { ...{ type: [String, Number] }, ...{
  required: true,
  default: () => [1, 2, 3],
} },
  \\"qux\\": { type: [String, Number] },
  \\"quux\\": { ...{ type: Boolean }, ...{ default: true } },
  \\"quuz\\": null,
})

import { computed as __MACROS_computed } from \\"vue\\";
import { expectTypeOf } from 'expect-type'
import type { ComputedRef } from 'vue'
import type { Qux } from './types'

const foo = __MACROS_computed(() => __MACROS_props[\\"foo\\"])
expectTypeOf(foo).toEqualTypeOf<ComputedRef<string>>()

const bar = __MACROS_computed(() => __MACROS_props[\\"bar\\"])
expectTypeOf(bar).toEqualTypeOf<ComputedRef<string>>()

const baz = __MACROS_computed(() => __MACROS_props[\\"baz\\"])
expectTypeOf(baz).toEqualTypeOf<ComputedRef<string | number>>()

__MACROS_computed(() => __MACROS_props[\\"qux\\"])

__MACROS_computed(() => __MACROS_props[\\"quux\\"])

// infer prop name from variable name
const quuz = __MACROS_computed(() => __MACROS_props[\\"quuz\\"])
console.log(quuz)
</script>
\`;

export { defineProp as default };
"
`;
